mesonbuild/interpreterbase.py

```diff
@@ -369,14 +369,19 @@ def evaluate_plusassign(self, node):
     def evaluate_indexing(self, node):
         assert(isinstance(node, mparser.IndexNode))
         iobject = self.evaluate_statement(node.iobject)
-        if not isinstance(iobject, list):
-            raise InterpreterException('Tried to index a non-array object.')
+        if not (isinstance(iobject, list) or hasattr(iobject, 'index_protocol')):
+            raise InterpreterException(
+                'Tried to index an object that doesn\'t support indexing.')
         index = self.evaluate_statement(node.index)
         if not isinstance(index, int):
             raise InterpreterException('Index value is not an integer.')
-        if index < -len(iobject) or index >= len(iobject):
+        try:
+            if isinstance(iobject, list):
+                return iobject[index]
+            else:
+                return iobject.index_protocol([index], {})
```

Comment:
... and this special case will go away.

Reply:
On the other hand, I really don't think that allowing python's list implementation to be directly exposed in meson is good, it's a layering violation. Wouldn't it be better to have a meson `List` and `ListHolder` object so that we don't rely on python implementation details?

Reply:
Yes, it is a layering violation, but we do such things in numerous places because there are advantages to that, namely:

1. Our corner-case behaviour is sane and 'non-surprising'/'expected' by default without too much work by us
2. The amount of work needed to implement features is a lot less
3. Parser behavior is well-documented because it is often the same as Python
4. Parsing is faster because directly using python builtins is faster than pure-python code
5. Porting Meson to another language is actually easier, because of (3)

We do have to be careful to not expose strange edge-cases that are Python-specific which may tie us to the language, but I don't think this is one of them.